File: libs/java/src/main/java/com/mbreissi/edgecommons/heartbeat/InstanceConnectivity.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.edgecommons.heartbeat;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * One component instance's southbound/source connectivity, reported at the <b>instance level</b>
 * through the component's {@code main} {@code state} keepalive (see
 * {@link InstanceConnectivityProvider}). The UNS model keeps a component's identity, data and
 * lifecycle under its {@code main} instance token — a multi-connection component (e.g. an OPC UA
 * adapter with several servers, a Modbus adapter with several slaves, a file-replicator with several
 * source directories) therefore does NOT mint a separate UNS instance per connection. Instead it
 * reports each connection's health here, and the console renders it per-instance under the one
 * component — no phantom "instance" component that never heartbeats.
 *
 * <p>The same sample answers both surfaces: it is pushed on every {@code state} keepalive tick, and
 * it is what the built-in {@code status} command verb returns when pulled
 * ({@link com.mbreissi.edgecommons.commands.CommandInbox#STATUS}). A component supplies the data
 * once; the library serves it both ways.
 *
 * <h2>The shape</h2>
 * <ul>
 *   <li>{@code instance} — the connection id. Required.</li>
 *   <li>{@code connected} — the <b>normalized</b> reachability flag. Always present, so a console can
 *       render a health dot for any component without knowing that component's vocabulary.</li>
 *   <li>{@code state} — optional, the component's <b>own</b> vocabulary for the richer condition
 *       ({@code ONLINE} / {@code CONNECTING} / {@code BACKOFF} / {@code DISABLED} …). A boolean cannot
 *       distinguish "reconnecting" from "administratively disabled"; this can.</li>
 *   <li>{@code detail} — optional human text (the endpoint, or why it is down).</li>
 *   <li>{@code attributes} — optional open bag of domain data (a camera's capabilities and last error,
 *       an OPC UA server's session id …). Deliberately unconstrained: it is where a component puts
 *       what only it understands, <b>without</b> destabilizing the fields above that everyone relies on.</li>
 * </ul>
 *
 * <p>Serialized into the state body's {@code instances[]} array, and into the {@code status} reply.
 */
public final class InstanceConnectivity
{
    private final String instance;
    private final boolean connected;
    private final String state;
    private final String detail;
    private final Map<String, JsonElement> attributes;

    /**
     * The private wide constructor. It is NOT public, and that is the point.
     *
     * <p>A public 5-arg positional constructor is what forced this type into a constructor-overload
     * and argument-ordering problem in the first place — one that had to be resolved differently in
     * each language (Python and TypeScript cannot overload a constructor at all). The public
     * construction path is therefore uniform across all four ports and positional-free:
     *
     * <pre>{@code
     * InstanceConnectivity.of("cam-02", false, "connect timed out")
     *     .withState("BACKOFF")
     *     .withAttributes(Map.of("lastError", new JsonPrimitive("CAMERA_UNAVAILABLE")));
     * }</pre>
     *
     * <p>which is exactly how it reads in Rust, TypeScript and Python. Adding a member later extends
     * the builder rather than minting a sixth argument nobody can order correctly. The legacy 3-arg
     * constructor stays public for back-compat.
     */
    private InstanceConnectivity(String instance, boolean connected, String detail, String state,
                                Map<String, JsonElement> attributes)
    {
        if (instance == null || instance.isBlank())
        {
            throw new IllegalArgumentException("instance id must be non-null and non-blank");
        }
        this.instance = instance;
        this.connected = connected;
        this.state = state;
        this.detail = detail;
        this.attributes = attributes == null || attributes.isEmpty()
                ? Map.of()
                : Map.copyOf(new LinkedHashMap<>(attributes));
    }

    /** Retained: the pre-{@code state}/{@code attributes} surface. */
    public InstanceConnectivity(String instance, boolean connected, String detail)
    {
        this(instance, connected, detail, null, null);
    }

    /** Convenience factory without a detail. */
    public static InstanceConnectivity of(String instance, boolean connected)
    {
        return new InstanceConnectivity(instance, connected, null, null, null);
    }

    /** Convenience factory with a detail. */
    public static InstanceConnectivity of(String instance, boolean connected, String detail)
    {
        return new InstanceConnectivity(instance, connected, detail, null, null);
    }

    /** Returns a copy carrying the component's own condition token. */
    public InstanceConnectivity withState(String newState)
    {
        return new InstanceConnectivity(instance, connected, detail, newState, attributes);
    }

    /** Returns a copy carrying domain-specific attributes. */
    public InstanceConnectivity withAttributes(Map<String, JsonElement> newAttributes)
    {
        return new InstanceConnectivity(instance, connected, detail, state, newAttributes);
    }

    public String getInstance()
    {
        return instance;
    }

    public boolean isConnected()
    {
        return connected;
    }

    /** The component's own richer condition token, or {@code null}. */
    public String getState()
    {
        return state;
    }

    /** The optional human detail (endpoint / down reason), or {@code null}. */
    public String getDetail()
    {
        return detail;
    }

    /** The domain-specific attributes; never {@code null}, possibly empty. */
    public Map<String, JsonElement> getAttributes()
    {
        return attributes;
    }

    /**
     * The wire element:
     * {@code {"instance":…,"connected":…[,"state":…][,"detail":…][,"attributes":{…}]}}.
     * Optional members are omitted rather than emitted null, so the common two-field case stays small
     * on a keepalive that ships every 5 seconds per component.
     */
    public JsonObject toJson()
    {
        JsonObject o = new JsonObject();
        o.addProperty("instance", instance);
        o.addProperty("connected", connected);
        if (state != null && !state.isBlank())
        {
            o.addProperty("state", state);
        }
        if (detail != null && !detail.isBlank())
        {
            o.addProperty("detail", detail);
        }
        if (!attributes.isEmpty())
        {
            JsonObject bag = new JsonObject();
            attributes.forEach(bag::add);
            o.add("attributes", bag);
        }
        return o;
    }
}

File: libs/python/edgecommons/heartbeat/instance_connectivity.py
"""Per-instance connectivity reporting for the ``state`` keepalive and the ``status``
verb (the #1c model).

A multi-connection component (an OPC UA adapter with several servers, a Modbus adapter
with several slaves, a file-replicator with several source directories) keeps its
identity, data and lifecycle at **component scope** (no instance token — D-U28) — it
does NOT mint a separate UNS instance per connection. Instead it reports each
connection's health here, and the console renders it per-instance under the one
component (no phantom "instance" component that never heartbeats).

Register a provider with ``gg.set_instance_connectivity_provider(...)``; the component's
``state`` keepalive then carries an ``instances`` array of :class:`InstanceConnectivity`.
The same sample answers both surfaces: it is pushed on every ``state`` keepalive tick,
and it is what the built-in ``status`` command verb
(:data:`edgecommons.command_inbox.STATUS`) returns when pulled. A component supplies the
data once; the library serves it both ways.

Mirrors ``libs/java/.../heartbeat/InstanceConnectivity.java`` (the Java canonical).
"""

from dataclasses import dataclass, replace
from typing import Any, Callable, Dict, List, Mapping, Optional


@dataclass(frozen=True)
class InstanceConnectivity:
    """One component instance's southbound/source connectivity.

    :param instance: the component instance / connection id (OPC UA server id, Modbus
        slave id, replication instance id); must be non-empty.
    :param connected: whether that instance's southbound/source is currently reachable —
        the **normalized** flag every consumer can read, so a console can render a health
        dot for any component without knowing that component's vocabulary.
    :param detail: an optional human detail (endpoint, or the down reason).
    :param state: optional — the component's **own** vocabulary for the richer condition
        (``ONLINE`` / ``CONNECTING`` / ``BACKOFF`` / ``DISABLED`` …). A boolean cannot
        distinguish "reconnecting" from "administratively disabled"; this can.
    :param attributes: optional open bag of domain data (a camera's capabilities and last
        error, an OPC UA server's session id …). Deliberately unconstrained: it is where a
        component puts what only it understands, **without** destabilizing the fields above
        that everyone relies on. Copied defensively.
    """

    instance: str
    connected: bool
    detail: Optional[str] = None
    state: Optional[str] = None
    attributes: Optional[Dict[str, Any]] = None

    def __post_init__(self) -> None:
        if not self.instance or not self.instance.strip():
            raise ValueError("instance id must be non-empty")
        if self.attributes is not None:
            if not isinstance(self.attributes, Mapping):
                raise ValueError("attributes must be a mapping")
            # Defensive copy: a later mutation of the caller's dict must not leak in.
            object.__setattr__(self, "attributes", dict(self.attributes))

    @staticmethod
    def of(instance: str, connected: bool, detail: Optional[str] = None) -> "InstanceConnectivity":
        """Convenience factory (the pre-``state``/``attributes`` surface, retained)."""
        return InstanceConnectivity(instance, connected, detail)

    def with_state(self, state: Optional[str]) -> "InstanceConnectivity":
        """A copy carrying the component's own condition token."""
        return replace(self, state=state)

    def with_attributes(self, attributes: Optional[Mapping[str, Any]]) -> "InstanceConnectivity":
        """A copy carrying domain-specific attributes."""
        return replace(self, attributes=dict(attributes) if attributes is not None else None)

    def to_dict(self) -> Dict[str, Any]:
        """The wire element:
        ``{"instance": …, "connected": …[, "state": …][, "detail": …][, "attributes": {…}]}``.

        Optional members are omitted rather than emitted null, so the common two-field case
        stays small on a keepalive that ships every 5 seconds per component.
        """
        d: Dict[str, Any] = {"instance": self.instance, "connected": self.connected}
        if self.state and self.state.strip():
            d["state"] = self.state
        if self.detail and self.detail.strip():
            d["detail"] = self.detail
        if self.attributes:
            d["attributes"] = dict(self.attributes)
        return d


# A component-supplied source of per-instance connectivity, sampled each keepalive tick
# and by the built-in ``status`` verb. A zero-arg callable returning the current
# per-instance connectivity; ``None``/empty omits the ``instances[]`` section. Must be
# cheap and non-blocking (sample a cached status) and should not raise (the heartbeat's
# sampling seam treats a raising provider as "no data this sample").
InstanceConnectivityProvider = Callable[[], Optional[List[InstanceConnectivity]]]
